pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -51,107 +53,142 @@ type kubeController struct {
+	serviceController *aggregate.Controller,
+	serviceEntryStore *serviceentry.ServiceEntryStore,
+	caBundlePath string,
+	fetchCaRoot func() map[string]string,
+	networksWatcher mesh.NetworksWatcher,
+) *Multicluster {
 	remoteKubeController := make(map[string]*kubeController)
 	if opts.ResyncPeriod == 0 {
 		// make sure a resync time of 0 wasn't passed in.
 		opts.ResyncPeriod = 30 * time.Second
 		log.Info("Resync time was configured to 0, resetting to 30")
 	}
 	mc := &Multicluster{
-		WatchedNamespaces:     opts.WatchedNamespaces,
-		DomainSuffix:          opts.DomainSuffix,
-		ResyncPeriod:          opts.ResyncPeriod,
+		serverID:              serverID,
+		opts:                  opts,
 		serviceController:     serviceController,
-		XDSUpdater:            xds,
+		serviceEntryStore:     serviceEntryStore,
+		caBundlePath:          caBundlePath,
+		fetchCaRoot:           fetchCaRoot,
+		XDSUpdater:            opts.XDSUpdater,
 		remoteKubeControllers: remoteKubeController,
 		networksWatcher:       networksWatcher,
-		metrics:               opts.Metrics,
-		fetchCaRoot:           opts.FetchCaRoot,
-		caBundlePath:          opts.CABundlePath,
-		systemNamespace:       opts.SystemNamespace,
 		secretNamespace:       secretNamespace,
-		endpointMode:          opts.EndpointMode,
 		syncInterval:          opts.GetSyncInterval(),
+		client:                kc,
 	}
-	mc.initSecretController(kc)
 
-	return mc, nil
+	return mc
 }
 
 // AddMemberCluster is passed to the secret controller as a callback to be called
 // when a remote cluster is added.  This function needs to set up all the handlers
 // to watch for resources being added, deleted or changed on remote clusters.
-func (m *Multicluster) AddMemberCluster(clients kubelib.Client, clusterID string) error {
+func (m *Multicluster) AddMemberCluster(client kubelib.Client, clusterID string) error {
 	// stopCh to stop controller created here when cluster removed.
 	stopCh := make(chan struct{})
 	var remoteKubeController kubeController
 	remoteKubeController.stopCh = stopCh
 	m.m.Lock()
-	options := Options{
-		SystemNamespace:   m.systemNamespace,
-		WatchedNamespaces: m.WatchedNamespaces,
-		ResyncPeriod:      m.ResyncPeriod,
-		DomainSuffix:      m.DomainSuffix,
-		XDSUpdater:        m.XDSUpdater,
-		ClusterID:         clusterID,
-		NetworksWatcher:   m.networksWatcher,
-		Metrics:           m.metrics,
-		EndpointMode:      m.endpointMode,
-		SyncInterval:      m.syncInterval,
-	}
+	options := m.opts
+	options.ClusterID = clusterID
 	log.Infof("Initializing Kubernetes service registry %q", options.ClusterID)
-	kubectl := NewController(clients, options)
+	kubeRegistry := NewController(client, options)
 
-	remoteKubeController.Controller = kubectl
-	m.serviceController.AddRegistry(kubectl)
+	remoteKubeController.Controller = kubeRegistry
+	m.serviceController.AddRegistry(kubeRegistry)
 
 	m.remoteKubeControllers[clusterID] = &remoteKubeController
+	firstCluster := len(m.remoteKubeControllers) == 1
```

Comment:
Compare `clusterID` to determine if this is primary cluster. We have passed the Options when constructing Multicluster